FLT/DivisionAlgebra/Finiteness.lean

```diff
@@ -118,19 +120,220 @@ lemma X_meets_kernel' {β : D_𝔸ˣ} (hβ : β ∈ ringHaarChar_ker D_𝔸) :
 /-- An auxiliary set T used in the proof of Fukisaki's lemma. Defined as Y ∩ Dˣ. -/
 def T : Set D_𝔸ˣ := ((↑) : D_𝔸ˣ → D_𝔸) ⁻¹' (Y K D) ∩ Set.range ((incl K D : Dˣ → D_𝔸ˣ))
 
--- Need something saying D ⊆ D_𝔸 is discrete
+/-- The K-algebra equivalence of D and K^n. -/
+abbrev D_iso : (D ≃ₗ[K] ((Fin (Module.finrank K D) → K))) := by
+  exact Module.Finite.equivPi K D
+
+/-- The 𝔸-algebra equivalence of D_𝔸 and 𝔸^d. -/
+def D𝔸_iso : (D_𝔸 ≃ₗ[(AdeleRing (𝓞 K) K)] ((Fin (Module.finrank K D) → AdeleRing (𝓞 K) K))) := by
+  suffices h : ((Fin (Module.finrank K D) → K) ⊗[K] AdeleRing (𝓞 K) K) ≃ₗ[(AdeleRing (𝓞 K) K)]
+      (Fin (Module.finrank K D) → AdeleRing (𝓞 K) K) by
+    have H1 := TensorProduct.AlgebraTensorModule.finiteEquivPi (R := K) (M := D)
+      (N := AdeleRing (𝓞 K) K)
+    have H2 : D ⊗[K] AdeleRing (𝓞 K) K ≃ₗ[AdeleRing (𝓞 K) K] (AdeleRing (𝓞 K) K) ⊗[K] D :=
+      (TensorProduct.RightActions.Module.TensorProduct.comm _ _ _).symm
+    exact H2.trans H1
+  have h1 := (TensorProduct.piScalarRight K (AdeleRing (𝓞 K) K) (AdeleRing (𝓞 K) K)
+    (Fin (Module.finrank K D)))
+  have h2 : (Fin (Module.finrank K D) → K) ⊗[K] AdeleRing (𝓞 K) K ≃ₗ[(AdeleRing (𝓞 K) K)]
+      AdeleRing (𝓞 K) K ⊗[K] (Fin (Module.finrank K D) → K) := by
+    exact (TensorProduct.RightActions.Module.TensorProduct.comm _ _ _).symm
+  exact h2.trans h1
+
+local instance : IsModuleTopology (AdeleRing (𝓞 K) K)
+    ((Fin (Module.finrank K D) → AdeleRing (𝓞 K) K)) := by
+  --have := IsModuleTopology.instPi (R := AdeleRing (𝓞 K) K) (ι := Fin (Module.finrank K D))
+  --  (A := Fin (Module.finrank K D) → AdeleRing (𝓞 K) K)
+
+    -- no idea how to get this to work
+  sorry
```

Comment:
This looks to me like a diamond in FLT. Right now it is unclear to me how to fix this but I'm actively thinking about it.